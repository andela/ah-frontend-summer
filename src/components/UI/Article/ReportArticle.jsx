import React from 'react';
import { connect } from 'react-redux';
import {
    Container,
    Segment,
    Form,
    Button,
    Radio,
    GridColumn,
    GridRow,
    Grid
} from 'semantic-ui-react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export const ReportArticle = props => {
    const { onSubmitHandler, onChangeHandler, slug, reason, loading } = props;
    const content = (
        <GridColumn width={8}>
            <Form onSubmit={onSubmitHandler}>
                <h3>Report Article</h3>
                <Form.Field>
                    <Radio
                        label='Spam'
                        name='spam'
                        value='spam'
                        checked={reason === 'spam'}
                        onChange={onChangeHandler}
                    />
                </Form.Field>
                <Form.Field>
                    <Radio
                        label='Harassment'
                        name='harassment'
                        value='harassment'
                        checked={reason === 'harassment'}
                        onChange={onChangeHandler}
                    />
                </Form.Field>
                <Form.Field>
                    <Radio
                        label='Rules Violation'
                        name='rules violation'
                        value='rules violation'
                        checked={reason === 'rules violation'}
                        onChange={onChangeHandler}
                    />
                </Form.Field>
                <Button basic color='teal' type='submit' loading={loading}>
                    Report
                </Button>
                <Link to={`/articles/${slug}`}>
                    <Button basic>Cancel</Button>
                </Link>
            </Form>
        </GridColumn>
    );

    return (
        <Container>
            <Grid>
                <GridRow>
                    <GridColumn width={4}></GridColumn>
                    <GridColumn width={8}>
                        <Segment padded='very' color='teal'>
                            <Grid>
                                <GridRow>
                                    <GridColumn width={4}></GridColumn>
                                    {content}
                                    <GridColumn width={4}></GridColumn>
                                </GridRow>
                            </Grid>
                        </Segment>
                    </GridColumn>
                    <GridColumn width={4}></GridColumn>
                </GridRow>
            </Grid>
        </Container>
    );
};

ReportArticle.propTypes = {
    onSubmitHandler: PropTypes.func.isRequired,
    onChangeHandler: PropTypes.func.isRequired,
    slug: PropTypes.string.isRequired,
    reason: PropTypes.string.isRequired
};

const mapStateToProps = state => {
    return {
        loading: state.reportArticle.loading
    };
};

export default connect(mapStateToProps)(ReportArticle);
