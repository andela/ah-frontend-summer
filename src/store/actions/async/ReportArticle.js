import axios from 'axios';
import { 
    reportArticleStarted, 
    reportArticleSucceded, 
    reportArticleFailed
} from '../sync/ReportArticle';

export const reportArticle = payload => {
    return async dispatch => {
        dispatch(reportArticleStarted());
        const { token, data, slug, push } = payload;
        try {
            await axios.post(
                `https://ah-backend-summer-staging.herokuapp.com/api/v1/articles/${slug}/report`,
                data,
                { headers: { Authorization: `Bearer ${token}` }}
            );
            dispatch(reportArticleSucceded('Article has been reported'));
            push(`/articles/${slug}`);
        } catch (error) {
            const errorMsg = 'Something went wrong, please check your internet connection';
            dispatch(reportArticleFailed(errorMsg));
        }
    };
};
