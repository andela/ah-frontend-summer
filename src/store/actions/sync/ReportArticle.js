import { 
    REPORT_ARTICLE_STARTED, 
    REPORT_ARTICLE_SUCCEDED, 
    REPORT_ARTICLE_FAILED,
    DISMISS_MESSAGE
} from "../actionTypes";

export const reportArticleStarted = () => {
    return {
        type: REPORT_ARTICLE_STARTED
    };
};

export const reportArticleSucceded = message => {
    return {
        type: REPORT_ARTICLE_SUCCEDED,
        message
    };
};

export const reportArticleFailed = errorMsg => {
    return {
        type: REPORT_ARTICLE_FAILED,
        errorMsg
    };
};

export const dismissMessage = () => {
    return {
        type: DISMISS_MESSAGE
    };
};
