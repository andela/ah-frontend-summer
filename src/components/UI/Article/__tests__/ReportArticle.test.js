import React from 'react';
import { shallow } from 'enzyme';
import { ReportArticle } from '../ReportArticle';

describe('Report Article', () => {
    const onSubmitHandlerMock = jest.fn();
    const onChangeHandlerMock = jest.fn();
    const props = {
        onSubmitHandler: onSubmitHandlerMock,
        onChangeHandler: onChangeHandlerMock,
        slug: "aska",
        reason: ''
    }
    const wrapper = shallow(<ReportArticle {...props}/>);

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('calls onSubmitHandler on form submission', () => {
        wrapper.find('Form').simulate('submit');
        expect(onSubmitHandlerMock).toHaveBeenCalled();
    });

    it('calls onChangeHandler when a radio button is checked', () => {
        wrapper.find('Radio[name="spam"]').simulate('change');
        expect(onChangeHandlerMock).toHaveBeenCalled();
    });
});
