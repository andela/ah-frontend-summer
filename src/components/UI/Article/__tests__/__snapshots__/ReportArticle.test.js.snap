// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Article renders correctly 1`] = `
<Container>
  <Grid>
    <GridRow>
      <GridColumn
        width={4}
      />
      <GridColumn
        width={8}
      >
        <Segment
          color="teal"
          padded="very"
        >
          <Grid>
            <GridRow>
              <GridColumn
                width={4}
              />
              <GridColumn
                width={8}
              >
                <Form
                  as="form"
                  onSubmit={[MockFunction]}
                >
                  <h3>
                    Report Article
                  </h3>
                  <FormField>
                    <Radio
                      checked={false}
                      label="Spam"
                      name="spam"
                      onChange={[MockFunction]}
                      type="radio"
                      value="spam"
                    />
                  </FormField>
                  <FormField>
                    <Radio
                      checked={false}
                      label="Harassment"
                      name="harassment"
                      onChange={[MockFunction]}
                      type="radio"
                      value="harassment"
                    />
                  </FormField>
                  <FormField>
                    <Radio
                      checked={false}
                      label="Rules Violation"
                      name="rules violation"
                      onChange={[MockFunction]}
                      type="radio"
                      value="rules violation"
                    />
                  </FormField>
                  <Button
                    as="button"
                    basic={true}
                    color="teal"
                    type="submit"
                  >
                    Report
                  </Button>
                  <Link
                    to="/articles/aska"
                  >
                    <Button
                      as="button"
                      basic={true}
                    >
                      Cancel
                    </Button>
                  </Link>
                </Form>
              </GridColumn>
              <GridColumn
                width={4}
              />
            </GridRow>
          </Grid>
        </Segment>
      </GridColumn>
      <GridColumn
        width={4}
      />
    </GridRow>
  </Grid>
</Container>
`;
